SPECS/filesystem/filesystem.spec

```diff
@@ -560,6 +560,24 @@ posix.mkdir("/proc")
 posix.mkdir("/sys")
 posix.chmod("/proc", 0555)
 posix.chmod("/sys", 0555)
+
+-- Prior to filesystem-1.1-16, /media used to be a symlink to /run/media but this was
+-- replaced with a directory. The RPM upgrade operation generally worked when the /media
+-- symlink is a dangling link, which is commonly the case, however not always the case.
+--
+-- And when the /media symlink is indeed properly pointing to a real /run/media, RPM has a
+-- known limitation where it is not possible to replace an active symlink with a directory,
+-- and thus the RPM transation fails.
+--
+-- To workaround this, a %pretrans scriptlet must run to test and remove the symlink
+-- before RPM attempts to install the new directory.
+--
+-- https://docs.fedoraproject.org/en-US/packaging-guidelines/Directory_Replacement
+path = "/media"
+st = posix.stat(path)
+if st and st.type == "link" then
```

Comment:
Do we need to consider the reverse, ie downgrading to the old version? [Docs have it here](https://docs.fedoraproject.org/en-US/packaging-guidelines/Directory_Replacement/#_scriptlet_to_replace_a_directory)

Or do we just consider that unsupported?

Reply:
@dmcilvaney - Looked at this closer and I do not think we can support the reverse operation, unless there is a better way to do this operation that is not risky / extremely brittle for the stable release. If no better way, I think we will be forced to consider the downgrade operation as unsupported, and need to communicate that appropriately.

Details:
The downgrade operation still follows the scriptlet ordering - https://docs.fedoraproject.org/en-US/packaging-guidelines/Scriptlets/#ordering

On a downgrade operation (i.e., to `filesystem-1.1-15` which has the /media symlink), the 1.1-15 RPM's `%pretrans` is first run, which will not have the requisite scriptlet to convert a /media directory back to a symlink. Then the installation progresses and hits the package installation conflict, which halts the transaction.

``` bash
$ sudo dnf --rpmverbosity=debug downgrade -y filesystem-1.1-15.cm2.x86_64
Last metadata expiration check: 1 day, 1:29:54 ago on Fri Dec  8 12:00:57 2023.
Dependencies resolved.
=============================================================================================================================================================================================================================================================================================================================
 Package                                                                    Architecture                                                           Version                                                                       Repository                                                                             Size
=============================================================================================================================================================================================================================================================================================================================
Downgrading:
 filesystem                                                                 x86_64                                                                 1.1-15.cm2                                                                    mariner-official-base                                                                  24 k

Transaction Summary
=============================================================================================================================================================================================================================================================================================================================
Downgrade  1 Package

Total size: 24 k
Downloading Packages:
[SKIPPED] filesystem-1.1-15.cm2.x86_64.rpm: Already downloaded
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Running scriptlet: filesystem-1.1-15.cm2.x86_64                                                                                                                                                                                                                                                                        1/1
D: PRAGMA secure_delete = OFF: 0
D: PRAGMA case_sensitive_like = ON: 0
D: PRAGMA journal_mode = WAL: 0
D: PRAGMA wal_autocheckpoint = 10000: 0
D: CREATE INDEX IF NOT EXISTS 'Name_key_idx' ON 'Name'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Basenames_key_idx' ON 'Basenames'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Basenames_hnum_idx' ON 'Basenames'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Group_key_idx' ON 'Group'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Requirename_key_idx' ON 'Requirename'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Requirename_hnum_idx' ON 'Requirename'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Providename_key_idx' ON 'Providename'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Providename_hnum_idx' ON 'Providename'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Conflictname_key_idx' ON 'Conflictname'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Conflictname_hnum_idx' ON 'Conflictname'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Obsoletename_key_idx' ON 'Obsoletename'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Obsoletename_hnum_idx' ON 'Obsoletename'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Triggername_key_idx' ON 'Triggername'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Triggername_hnum_idx' ON 'Triggername'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Dirnames_key_idx' ON 'Dirnames'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Dirnames_hnum_idx' ON 'Dirnames'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Sha1header_key_idx' ON 'Sha1header'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Filetriggername_key_idx' ON 'Filetriggername'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Filetriggername_hnum_idx' ON 'Filetriggername'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Transfiletriggername_key_idx' ON 'Transfiletriggername'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Transfiletriggername_hnum_idx' ON 'Transfiletriggername'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Recommendname_key_idx' ON 'Recommendname'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Recommendname_hnum_idx' ON 'Recommendname'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Suggestname_key_idx' ON 'Suggestname'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Suggestname_hnum_idx' ON 'Suggestname'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Supplementname_key_idx' ON 'Supplementname'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Supplementname_hnum_idx' ON 'Supplementname'(hnum ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Enhancename_key_idx' ON 'Enhancename'(key ASC): 0
D: CREATE INDEX IF NOT EXISTS 'Enhancename_hnum_idx' ON 'Enhancename'(hnum ASC): 0
D: sanity checking 2 elements
D:  read h#     451
Header SHA256 digest: OK
Header SHA1 digest: OK
D: Plugin: calling hook tsm_pre in selinux plugin
D: Plugin: calling hook tsm_pre in syslog plugin
D: running pre-transaction scripts
D: filesystem-1.1-15.cm2.x86_64: Header V4 RSA/SHA256 Signature, key ID 3135ce90: OK
D: filesystem-1.1-15.cm2.x86_64: Header SHA256 digest: OK
D: filesystem-1.1-15.cm2.x86_64: Header SHA1 digest: OK
D: Plugin: calling hook psm_pre in selinux plugin
D: %pretrans(filesystem-1.1-15.cm2.x86_64): running <lua> scriptlet.
D: Plugin: calling hook scriptlet_post in syslog plugin

  Preparing        :                                                                                                                                                                                                                                                                                                     1/1
Transaction couldn't start:
file /media from install of filesystem-1.1-15.cm2.x86_64 conflicts with file from package filesystem-1.1-19.cm2.x86_64
The downloaded packages were saved in cache until the next successful transaction.
You can remove cached packages by executing 'dnf clean packages'.
Error: Could not run transaction.

```

Reply:
For m2, this does look like the best approach, and also I agree there's no sane way to also get downgrading to work (i.e. dir->symlink).

For azl3, the scriptlet shouldnt' be needed since we're in the dev phase...and also, I think changes like this probably should only happen in the dev phase.

Reply:
Yup, azl3 should not have this scriptlet. It should only have the `/media` as a directory change, which is currently present in its dev branch.

Reply:
Is it worth putting a poison pill in the `%preun` of the new package? I think if that script returns an error it will abort the install of that package. Unfortunately, I can't find a good way to get the "new" version of a package being installed. Might just have to take the risk. I think this is likely not a common use-case, and downgrade will be even less common...

Reply:
`%preun` wouldn't work because that scriptlet doesn't run before the conflict is seen (see verbose output posted)
The only scriptlet that runs is the new package's `%pretrans` lua scriptlet and I really don't think we should add poison pill logic in there.

Note: we're still talking about erroring out the package downgrade, which is exactly what happens today.